File: crates/yauth/src/domain/oauth.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OauthAccount {
    pub id: Uuid,
    pub user_id: Uuid,
    pub provider: String,
    pub provider_user_id: String,
    pub access_token_enc: Option<String>,
    pub refresh_token_enc: Option<String>,
    pub created_at: NaiveDateTime,
    pub expires_at: Option<NaiveDateTime>,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NewOauthAccount {
    pub id: Uuid,
    pub user_id: Uuid,
    pub provider: String,
    pub provider_user_id: String,
    pub access_token_enc: Option<String>,
    pub refresh_token_enc: Option<String>,
    pub created_at: NaiveDateTime,
    pub expires_at: Option<NaiveDateTime>,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OauthState {
    pub state: String,
    pub provider: String,
    pub redirect_url: Option<String>,
    pub expires_at: NaiveDateTime,
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NewOauthState {
    pub state: String,
    pub provider: String,
    pub redirect_url: Option<String>,
    pub expires_at: NaiveDateTime,
    pub created_at: NaiveDateTime,
}

File: domain/oauth.go
package domain

import "time"

// OAuthAccount links a user to an external OAuth provider identity.
type OAuthAccount struct {
	ID              string
	UserID          string
	Provider        string
	ProviderUserID  string
	AccessTokenEnc  *string
	RefreshTokenEnc *string
	CreatedAt       time.Time
	ExpiresAt       *time.Time
	UpdatedAt       time.Time
}

// NewOAuthAccount is the input for creating an OAuth account link.
type NewOAuthAccount struct {
	ID              string
	UserID          string
	Provider        string
	ProviderUserID  string
	AccessTokenEnc  *string
	RefreshTokenEnc *string
	CreatedAt       time.Time
	ExpiresAt       *time.Time
	UpdatedAt       time.Time
}

// OAuthState is a CSRF-protecting state token for the OAuth authorization code
// flow.
type OAuthState struct {
	State       string
	Provider    string
	RedirectURL *string
	ExpiresAt   time.Time
	CreatedAt   time.Time
}

// NewOAuthState is the input for creating an OAuth state token.
type NewOAuthState struct {
	State       string
	Provider    string
	RedirectURL *string
	ExpiresAt   time.Time
	CreatedAt   time.Time
}
